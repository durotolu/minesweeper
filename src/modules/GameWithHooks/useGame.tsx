import React, { FC, useState, useEffect, useCallback } from "react";

import {
  CellState,
  Coords,
  Field,
  generateFieldWithDefaultState,
  fieldGenerator,
} from "@/core/Field";

import { LevelNames, GameSettings } from "@/modules/GameSettings";
import { openCell } from "@/core/openCell";
import { setFlag } from "@/core/setFlag";

import { useStatus } from "./useStatus";
import { useTime } from "./useTime";
import { useSettings } from "./useSettings";

interface ReturnType {
  level: LevelNames;
  time: number;
  isGameOver: boolean;
  isGameStarted: boolean;
  isWin: boolean;
  settings: [number, number];
  playerField: Field;
  gameField: Field;
  flagCounter: number;
  onClick: (coords: Coords) => void;
  onContextMenu: (coords: Coords) => void;
  onChangeLevel: (level: LevelNames) => void;
  onReset: () => void;
}

export const useGame = (
  defaultLevel = "beginner" as LevelNames
): ReturnType => {
  const {
    settings: [size, bombs],
    level,
    setLevel,
  } = useSettings(defaultLevel);

  const {
    isGameStarted,
    isGameOver,
    isWin,
    setNewGame,
    setInProgress,
    setGameWin,
    setGameLoose,
  } = useStatus();

  const [time, resetTime] = useTime(isGameStarted, isGameOver);
  const [flagCounter, setFlagCounter] = useState(0);

  const [playerField, setPlayerField] = useState<Field>(
    generateFieldWithDefaultState(size, CellState.hidden)
  );

  const [gameField, setGameField] = useState<Field>(
    fieldGenerator(size, bombs / (size * size))
  );

  const onClick = useCallback(
    (coords: Coords) => {
      !isGameStarted && setInProgress();
      try {
        const [newPlayerField, isSolved] = openCell(
          coords,
          playerField,
          gameField
        );
        if (isSolved) {
          setGameWin();
        }
        setPlayerField([...newPlayerField]);
      } catch (e) {
        setPlayerField([...gameField]);
        setGameLoose();
      }
    },
    [
      isGameStarted,
      isGameOver,
      isWin,
      level,
      flagCounter,
      playerField,
      gameField,
    ]
  );

  const onChangeLevel = useCallback((level: LevelNames) => {
    const newSettings = setLevel(level);
    resetHandler(newSettings);
  }, []);

  const onContextMenu = useCallback(
    (coords: Coords) => {
      !isGameStarted && setInProgress();
      const [newPlayerField, isSolved, newFlagCounter] = setFlag(
        coords,
        playerField,
        gameField,
        flagCounter,
        bombs
      );
      setFlagCounter(newFlagCounter);
      if (isSolved) {
        setGameWin();
      }
      setPlayerField([...newPlayerField]);
    },
    [
      isGameStarted,
      isGameOver,
      isWin,
      level,
      flagCounter,
      playerField,
      gameField,
    ]
  );

  const resetHandler = ([size, bombs]: [number, number]) => {
    const newGameField = fieldGenerator(size, bombs / (size * size));
    const newPlayerField = generateFieldWithDefaultState(
      size,
      CellState.hidden
    );
    setGameField([...newGameField]);
    setPlayerField([...newPlayerField]);
    setNewGame();
    resetTime();
    setFlagCounter(0);
  };

  const onReset = useCallback(() => resetHandler([size, bombs]), [size, bombs]);

  return {
    level,
    time,
    isGameOver,
    isGameStarted,
    isWin,
    settings: [size, bombs],
    playerField,
    gameField,
    flagCounter,
    onClick,
    onChangeLevel,
    onContextMenu,
    onReset,
  };
};
