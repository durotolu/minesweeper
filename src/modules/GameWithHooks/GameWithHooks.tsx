import React, { FC, useCallback } from "react";
import { useSearchParams } from "react-router-dom";
import { useGame } from "./useGame";

import { GameLevels, LevelNames } from "@/modules/GameSettings";

import { Grid } from "@/components/Grid";
import { Top } from "@/components/Top";
import { Scoreboard } from "@/components/Scoreboard";
import { GameArea, Wrapper, GameOver } from "@/components/Game";

export const GameWithHooks: FC = () => {
  const [searchParams, setSearchParams] = useSearchParams();
  const urlLevelParam = (searchParams.get("level") || undefined) as LevelNames;

  const {
    level,
    time,
    isGameOver,
    isWin,
    settings: [size, bombs],
    playerField,
    flagCounter,
    onClick,
    onChangeLevel,
    onContextMenu,
    onReset,
  } = useGame(urlLevelParam);

  const onChangeLevelHandler = useCallback(
    ({ target: { value: level } }: React.ChangeEvent<HTMLSelectElement>) => {
      setSearchParams({ level })
      onChangeLevel(level as LevelNames);
    },
    []
  );

  return (
    <Wrapper>
      <Top feature="Flag" firstAction="right click">
        Minesweeper
      </Top>
      <GameArea>
        <Scoreboard
          time={String(time)}
          bombs={String(bombs - flagCounter)}
          levels={GameLevels as unknown as LevelNames[]}
          defaultLevel={level}
          onReset={onReset}
          onChangeLevel={onChangeLevelHandler}
        />
        {isGameOver && <GameOver onClick={onReset} isWin={isWin} />}
        <Grid onClick={onClick} onContextMenu={onContextMenu}>
          {playerField}
        </Grid>
      </GameArea>
    </Wrapper>
  );
};
