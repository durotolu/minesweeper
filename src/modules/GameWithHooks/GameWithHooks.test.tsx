import React from "react";
import { MemoryRouter, BrowserRouter, useSearchParams } from "react-router-dom";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { GameWithHooks } from "./GameWithHooks";

const mockOnClick = jest.fn();
const mockOnChangeLevel = jest.fn();
const mockOnReset = jest.fn();
const mockOnContextMenu = jest.fn();
const mockSetSearchParams = jest.fn();

// jest.mock("react-router-dom", () => ({
//   useSearchParams: jest.fn,
// }));
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useSearchParams: jest.fn(),
}));

(useSearchParams as jest.Mock).mockReturnValue([
  { get: () => null },
  mockSetSearchParams,
]);

jest.mock("./useGame", () => ({
  __esModule: true,
  useGame: (level = "beginner") => ({
    level,
    time: 0,
    flagCounter: 0,
    isGameOver: true,
    isWin: false,
    settings: [9, 10],
    playerField: [
      [10, 10],
      [10, 10],
    ],
    onClick: mockOnClick,
    onContextMenu: mockOnContextMenu,
    onChangeLevel: mockOnChangeLevel,
    onReset: mockOnReset,
  }),
}));

beforeEach(() => {
  jest.clearAllMocks();
});

describe("GameWithHooks test cases", () => {
  describe("Render behaviour", () => {
    it("Render game field by default", () => {
      const { asFragment } = render(
        <MemoryRouter
          initialEntries={[{ pathname: "/", search: "?level=beginner" }]}
        >
          <GameWithHooks />
        </MemoryRouter>
      );
      expect(asFragment()).toMatchSnapshot();
    });
    it("Cell click works fine", async () => {
      render(
        <MemoryRouter
          initialEntries={[{ pathname: "/", search: "?level=beginner" }]}
        >
          <GameWithHooks />
        </MemoryRouter>
      );
      await userEvent.click(screen.getByTestId("0,0"));
      expect(mockOnClick).toHaveBeenCalled();
    });
    it("Context menu handler on a cell works fine", async () => {
      render(
        <MemoryRouter
          initialEntries={[{ pathname: "/", search: "?level=beginner" }]}
        >
          <GameWithHooks />
        </MemoryRouter>
      );
      await userEvent.pointer({
        keys: "[MouseRight>]",
        target: screen.getByTestId("0,0"),
      });
      expect(mockOnContextMenu).toHaveBeenCalled();
    });
    it("Reset handler works fine", async () => {
      render(
        <MemoryRouter
          initialEntries={[{ pathname: "/", search: "?level=beginner" }]}
        >
          <GameWithHooks />
        </MemoryRouter>
      );
      await userEvent.click(screen.getByRole("button"));
      expect(mockOnReset).toHaveBeenCalled();
    });
    it("Change level works fine", async () => {
      render(
        <MemoryRouter
          initialEntries={[{ pathname: "/", search: "?level=beginner" }]}
        >
          <GameWithHooks />
        </MemoryRouter>
      );
      await userEvent.selectOptions(
        screen.getByRole("combobox"),
        "intermediate"
      );
      expect(mockOnChangeLevel).toHaveBeenCalled();
      expect(mockSetSearchParams).toHaveBeenCalledWith({
        level: `intermediate`,
      });
    });
    it("Level in search params works fine", () => {
      (useSearchParams as jest.Mock).mockReturnValue([
        { get: () => "intermediate" },
        mockSetSearchParams,
      ]);

      render(
        <MemoryRouter
          initialEntries={[{ pathname: "/", search: "?level=intermediate" }]}
        >
          <GameWithHooks />
        </MemoryRouter>
      );
      const intermediateOption = screen.queryByText("intermediate");
      expect(intermediateOption).toBeInTheDocument();
    });
    it("Game over reset the game state", async () => {
      render(
        <MemoryRouter
          initialEntries={[{ pathname: "/", search: "?level=beginner" }]}
        >
          <GameWithHooks />
        </MemoryRouter>
      );
      await userEvent.click(screen.getByText("🙁"));
      expect(mockOnReset).toHaveBeenCalled();
    });
  });
});
