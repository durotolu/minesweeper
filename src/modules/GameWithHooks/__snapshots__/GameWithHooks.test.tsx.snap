// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameWithHooks test cases Render behaviour Render game field by default 1`] = `
<DocumentFragment>
  <div
    class="css-1ct4owg"
  >
    <header
      class="css-dn2347"
    >
      <h1
        class="css-18kvaxm"
      >
        Minesweeper
      </h1>
      <legend
        class="css-17qz6a2"
      >
        <strong>
          Flag: 
        </strong>
        <code
          class="css-12ap3z0"
        >
          <span
            class="css-udfepp"
          >
            right click
          </span>
           + 
          <span
            class="css-9c23k1"
          />
        </code>
      </legend>
    </header>
    <div
      class="css-1bh93n8"
    >
      <div
        class="css-1n7y7g1"
      >
        <div
          class="css-d2q1pa"
        >
          0
        </div>
        <div>
          <select
            class="css-1lpmfg0"
          >
            <option
              class="css-zcmdxr"
              selected=""
              value="beginner"
            >
              beginner
            </option>
            <option
              class="css-zcmdxr"
              value="intermediate"
            >
              intermediate
            </option>
            <option
              class="css-zcmdxr"
              value="expert"
            >
              expert
            </option>
          </select>
          <button
            class="css-1kb3z2h"
          >
            🙂
          </button>
        </div>
        <div
          class="css-d2q1pa"
        >
          10
        </div>
      </div>
      <div
        class="css-1p4hvro"
      >
        🙁
      </div>
      <div
        class="css-da0epy"
        role="grid"
        size="2"
      >
        <div
          class="css-1tt7j02"
          data-testid="0,0"
          role="cell"
        >
          10
        </div>
        <div
          class="css-1tt7j02"
          data-testid="0,1"
          role="cell"
        >
          10
        </div>
        <div
          class="css-1tt7j02"
          data-testid="1,0"
          role="cell"
        >
          10
        </div>
        <div
          class="css-1tt7j02"
          data-testid="1,1"
          role="cell"
        >
          10
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
